network/common/src/main/java/org/apache/spark/network/buffer/WrappedLargeByteBuffer.java

```diff
@@ -0,0 +1,252 @@
+      b.position(0);
+      dataCopy[i].put(b);
+      dataCopy[i].position(0);
+      b.position(originalPosition);
+    }
+    return new WrappedLargeByteBuffer(dataCopy);
+  }
+
+  @Override
+  public byte get() {
+    byte r = currentBuffer.get();
+    _pos += 1;
+    updateCurrentBuffer();
+    return r;
+  }
+
+  private void updateCurrentBuffer() {
+    while (currentBuffer != null && !currentBuffer.hasRemaining()) {
+      currentBufferIdx += 1;
+      currentBuffer = currentBufferIdx < underlying.length ? underlying[currentBufferIdx] : null;
+    }
+  }
+
+  @Override
+  public long position() {
+    return _pos;
+  }
+
+  @Override
+  public long skip(long n) {
+    if (n < 0) {
+      final long moveTotal = Math.min(-n, _pos);
+      long toMove = moveTotal;
+      // move backwards -- set the position to 0 of every buffer's we go back
+      if (currentBuffer != null) {
+        currentBufferIdx += 1;
+      }
+      while (toMove > 0) {
+        currentBufferIdx -= 1;
+        currentBuffer = underlying[currentBufferIdx];
+        int thisMove = (int) Math.min(toMove, currentBuffer.position());
+        currentBuffer.position(currentBuffer.position() - thisMove);
+        toMove -= thisMove;
+      }
+      _pos -= moveTotal;
+      return -moveTotal;
+    } else if (n > 0) {
+      final long moveTotal = Math.min(n, remaining());
+      long toMove = moveTotal;
+      // move forwards-- set the position to the end of every buffer as we go forwards
+      currentBufferIdx -= 1;
+      while (toMove > 0) {
+        currentBufferIdx += 1;
+        currentBuffer = underlying[currentBufferIdx];
+        int thisMove = (int) Math.min(toMove, currentBuffer.remaining());
+        currentBuffer.position(currentBuffer.position() + thisMove);
+        toMove -= thisMove;
+      }
+      _pos += moveTotal;
+      return moveTotal;
+    } else {
+      return 0;
+    }
+  }
+
+  @Override
+  public long remaining() {
+    return size - _pos;
+  }
+
+  @Override
+  public WrappedLargeByteBuffer duplicate() {
+    ByteBuffer[] duplicates = new ByteBuffer[underlying.length];
+    for (int i = 0; i < underlying.length; i++) {
+      duplicates[i] = underlying[i].duplicate();
+    }
+    return new WrappedLargeByteBuffer(duplicates, _pos);
+  }
+
+  @Override
+  public long size() {
+    return size;
+  }
+
+  @Override
+  public long writeTo(WritableByteChannel channel) throws IOException {
+    long written = 0l;
+    for (; currentBufferIdx < underlying.length; currentBufferIdx++) {
+      currentBuffer = underlying[currentBufferIdx];
+      written += currentBuffer.remaining();
+      while (currentBuffer.hasRemaining())
+        channel.write(currentBuffer);
+    }
+    _pos = size();
+    return written;
+  }
+
+  @Override
+  public ByteBuffer asByteBuffer() throws BufferTooLargeException {
+    if (underlying.length > 1) {
+      throw new BufferTooLargeException(size());
+    }
+    return underlying[0];
```

Comment:
This doesn't really follow the interface's contract.  E.g., it doesn't guarantee the buffer will have position 0; nor does it return as much data as possible (e.g. if the first buffer is less than the maximum size, you could include data from the second buffer, and so on).